### PR TITLE
Parameter and pinout cleanup #277

### PR DESCRIPTION
**Pull request by Silabs-ArjanB**
Signed-off-by: Arjan Bink <Arjan.Bink@silabs.com>

### CODE REVIEW

**Comment by Silabs-ArjanB** [comment]
Needs to be done at the same time as https://github.com/openhwgroup/core-v-verif/pull/85

**Comment by Silabs-ArjanB** [comment]
Added back in the APU interface.

**Comment by davideschiavone** [comment]
Thanks a lot @Silabs-ArjanB ,

few comments here:

1. The tb/dm testbench hasn't been updated with the new interface and parameters

2. Fixing PULP_SECURE to 0 makes the MTVEC(and MTVECx) tight to Boot Address, this makes the tests failing. It's a good example for issue #254
We need to fix #254 before having this PR merged. So I opened in parallel a PR for #254 , then you can merge it into your branch to have this point solved.

**Comment by Silabs-ArjanB** [comment]
The status of the tb/dm is unclear to me. They already seemed out of date to begin with. Can you let me know which ones are actually in use and I will update them (some of them seem to have been copied into the core-v-verif repos instead).

Do you want me to work on #254?

**Comment by davideschiavone** [comment]
PR for #254  is already here: https://github.com/openhwgroup/cv32e40p/pull/285
You just need to review.

tb/dm should just compile and work as before. The purpose is to provide an example to use GDB.
The `tb` folder will be replaced with the `example` name as its content is not for verification but just to provide quick examples. One with GDB/Debug and one with hello world.

**Comment by davideschiavone** [comment]
@Silabs-ArjanB , as the PR #285 has been merged to master, the point 2. should be solved.

>     2. Fixing PULP_SECURE to 0 makes the MTVEC(and MTVECx) tight to Boot Address, this makes the tests failing. It's a good example for issue #254
>        We need to fix #254 before having this PR merged. So I opened in parallel a PR for #254 , then you can merge it into your branch to have this point solved.

You should now merge the master to your branch and see if it solves the issue and address the point 1 here:

>     1. The tb/dm testbench hasn't been updated with the new interface and parameters

Thanks @Silabs-ArjanB

**Comment by Silabs-ArjanB** [comment]
Cleaned up parameters and pinout in various testbenches. 

- make vsim-run in 'cv32e40p/tb/dm' now compiles without errors and warnings. C code comilation however fails (I expect it was already like this)

- make custom-vsim-run in cv32e40p/tb/core now compiles without errors and warnings and executes okay as well.
- I don't have verilator so I couldn't check those compile targets (but updated all testbenches)
- make firmware-vsim-run-gui in cv32e40p/tb/core has gcc compilation issues (maybe a Makefile issue as 64-bit code is getting compiled it seems).

Could we open another ticket to get rid of the high amount of testbenches and duplication in cv32e40p. For example riscv_core is currently being instantiated in all of the following:

cv32e40p/tb/core/riscv_wrapper.sv
cv32e40p/tb/dm/tb_test_env.sv
cv32e40p/tb/dm/tb_top_verilator.sv
cv32e40p/tb/verilator-model/top.sv
cv32e40p/tb/tb_riscv/tb_riscv_core.sv

and riscv_wrapper for example is then being instantiated in 

tb/core/tb_top.sv 
tb/core/tb_top_verilator.sv

I updated all of the above but we should try to get rid of obsolete testbenches.

**Comment by davideschiavone** [comment]
Hi @Silabs-ArjanB , my answers inline
> 
> Cleaned up parameters and pinout in various testbenches.
> 
>     * make vsim-run in 'cv32e40p/tb/dm' now compiles without errors and warnings. C code comilation however fails (I expect it was already like this
It works for me

> 
>     * make custom-vsim-run in cv32e40p/tb/core now compiles without errors and warnings and executes okay as well.

👍 

> 
>     * I don't have verilator so I couldn't check those compile targets (but updated all testbenches)
> 
>     * make firmware-vsim-run-gui in cv32e40p/tb/core has gcc compilation issues (maybe a Makefile issue as 64-bit code is getting compiled it seems).

The problem is the RVA instructions emitted. As you tight to 0 the RVA parameter, you should also disable those instructions.
To do so, please remove them from firmware/start.S and from the Makefile variables
`RISCV_TEST_INCLUDES ` and `FIRMWARE_TEST_OBJS `

> 
> 
> Could we open another ticket to get rid of the high amount of testbenches and duplication in cv32e40p. For example riscv_core is currently being instantiated in all of the following:
> 
> cv32e40p/tb/core/riscv_wrapper.sv
> cv32e40p/tb/dm/tb_test_env.sv
> cv32e40p/tb/dm/tb_top_verilator.sv
> cv32e40p/tb/verilator-model/top.sv
> cv32e40p/tb/tb_riscv/tb_riscv_core.sv
> 
> and riscv_wrapper for example is then being instantiated in
> 
> tb/core/tb_top.sv
> tb/core/tb_top_verilator.sv
> 
> I updated all of the above but we should try to get rid of obsolete testbenches.
Yes please go ahead. We already have a ticket to change name to the folder and clean it up. @MikeOpenHWGroup is porting the interrupt test. Then, we should only have an example of application and the debug connected.

**Comment by Silabs-ArjanB** [comment]
Hi @davideschiavone , thank you for your feedback.

I will fix the RVA instruction issue as you described.

Can you be more specific about which parts I can remove please? For example, can I remove all of the following?

Files: 

cv32e40p/tb/dm/tb_top_verilator.sv
tb/core/tb_top_verilator.sv

Directories:

cv32e40p/tb/verilator-model
cv32e40p/tb/serDiv
cv32e40p/tb/tb_MPU
cv32e40p/tb/tb_riscv

**Comment by davideschiavone** [comment]
Hi @Silabs-ArjanB ,

Can we have that discussion in another thread/issue ?

File: rtl/riscv_core.sv [hunk 1]
```diff
@@ -75,11 +58,11 @@ module riscv_core
...
-  output logic                         instr_req_o,
-  input  logic                         instr_gnt_i,
-  input  logic                         instr_rvalid_i,
-  output logic                  [31:0] instr_addr_o,
-  input  logic [INSTR_RDATA_WIDTH-1:0] instr_rdata_i,
+  output logic        instr_req_o,
+  input  logic        instr_gnt_i,
+  input  logic        instr_rvalid_i,
+  output logic [31:0] instr_addr_o,
+  input  logic [31:0] instr_rdata_i,
```
Comment (MikeOpenHWGroup) [inline rtl/riscv_core.sv:65]: Why are we removing parameters here?  Even if these widths must be fixed, there are a _lot_ of magic numbers in the code and it would be a good idea to slowly replace them all.  How about:
```
output logic                          instr_req_o,
input  logic                          instr_gnt_i,
input  logic                          instr_rvalid_i,
output logic [INSTR_ADDR_WIDTH -1:0]  instr_addr_o,
input  logic [INSTR_RDATA_WIDTH-1:0]  instr_rdata_i,
```

File: rtl/riscv_core.sv [hunk 2]
```diff
@@ -91,8 +74,6 @@ module riscv_core
   output logic [31:0] data_wdata_o,
```
Comment (MikeOpenHWGroup) [inline rtl/riscv_core.sv:74]: More magic numbers.  Consider replacing with parameter, even if these are intended to be fixed values for cv32e.

File: rtl/riscv_core.sv [hunk 3]
```diff
@@ -476,7 +476,7 @@ module riscv_core
     .N_HWLP              ( N_HWLP            ),
     .RDATA_WIDTH         ( INSTR_RDATA_WIDTH ),
     .FPU                 ( FPU               ),
-    .DM_HaltAddress      ( DM_HaltAddress    )
+    .DM_HALTADDRESS      ( DM_HALTADDRESS    )
```
Comment (MikeOpenHWGroup) [inline rtl/riscv_core.sv:479]: Nice.  All caps for parameters is a good coding style.

File: rtl/riscv_core.sv [hunk 4]
```diff
@@ -561,7 +561,7 @@ module riscv_core
     .A_EXTENSION                  ( A_EXTENSION          ),
     .APU                          ( APU                  ),
     .FPU                          ( FPU                  ),
-    .Zfinx                        ( Zfinx                ),
+    .PULP_ZFINX                   ( PULP_ZFINX           ),
```
Comment (MikeOpenHWGroup) [inline rtl/riscv_core.sv:564]: Dito.

**Comment by Silabs-ArjanB** [comment]
> 
> 
> Hi @Silabs-ArjanB ,
> 
> Can we have that discussion in another thread/issue ?

Sure

File: rtl/riscv_core.sv (hunk 1 above)
Comment (Silabs-ArjanB) [reply rtl/riscv_core.sv:65]: Only instr_rdata_i had a parameter. There is no need to use a parameter for instr_addr_o. This will always be 32 bits wide. The use of INSTR_ADDR_WIDTH is debatable; I would not be against it but at the same time both cv32e40p and cv32e40 will have it set to 32.

File: rtl/riscv_core.sv (hunk 2 above)
Comment (Silabs-ArjanB) [reply rtl/riscv_core.sv:74]: It is a 32-bit core. Let's not go adding even more parameters

**Comment by Silabs-ArjanB** [comment]
Fixed Makefile and firmware/start.S to no longer use non-supported instructions (i.e. atomics). Note that riscv_tests/rv32uc/rvc.S needed to be changed (as was done before) as it led to an internal compiler error.

**Comment by davideschiavone** [comment]
For me it's ok to MERGE, just hitting @bluewww to ask him about the last change you have done in the RVC (he is the expert :) )

File: tb/core/firmware/start.S [hunk 5]
```diff
@@ -317,17 +317,6 @@ start:
...
-	TEST(amoand_w)
-	TEST(amomaxu_w)
-	TEST(amomax_w)
-	TEST(amominu_w)
-	TEST(amomin_w)
-	TEST(amoor_w)
-	TEST(amoswap_w)
-	TEST(amoxor_w)
-	TEST(lrsc)
-
```
Comment (bluewww) [inline tb/core/firmware/start.S:330]: Why is this removed? These test should work.

Comment (davideschiavone) [reply tb/core/firmware/start.S:330]: The RVA parameter has been set to 0 by default, so they generate illegal instructions now

File: tb/core/mm_ram.sv [hunk 6]
```diff
@@ -715,7 +716,7 @@ module mm_ram
     .req_core_i         ( instr_req_i            ),
     .req_mem_o          ( rnd_stall_instr_req    ),
 
-    .addr_core_i        ( 32'(instr_addr_i)      ),
```
Comment (bluewww) [inline tb/core/mm_ram.sv:718]: Removing this will generate a port mismatch warning

File: tb/core/riscv_tests/rv32uc/rvc.S [hunk 7]
```diff
@@ -1,7 +1,156 @@
...
+  .option pop
+
+  TEST_PASSFAIL
+
+RVTEST_CODE_END
+
+  .data
+RVTEST_DATA_BEGIN
 
-#include "../rv64uc/rvc.S"
```
Comment (bluewww) [inline tb/core/riscv_tests/rv32uc/rvc.S:7]: Is this a copy paste of the `../rv64uc/rvc.S` file? Why would you do this for this file, but then you don't do it for the others?

There is a very good reason the structure of the test files is like this (i.e. with the `#include`s from `rv64uc`). They are a verbatim import from https://github.com/riscv/riscv-tests meaning that whenever there is an update to the upstream test repository, you can easily import the diff. Now that you change the structure you have to apply fixes everytime you do this. This is just an unnecessary increase in maintenance cost.

All in all its inconsistent and a bad idea. I would highly suggest to revert this change and apply the fix to `../rv64uc/rvc.S` which I guess is removing the `.option norelax`, a fix I mistakenly did because I compiled with `-march=rv32im` instead of `imc` I suspect.

File: tb/core/riscv_wrapper.sv [hunk 8]
```diff
@@ -42,7 +43,7 @@ module riscv_wrapper
     logic [3:0]                   data_be;
     logic [31:0]                  data_rdata;
     logic [31:0]                  data_wdata;
-    logic [5:0]                   data_atop;
+    logic [5:0]                   data_atop = 'b0;
```
Comment (bluewww) [inline tb/core/riscv_wrapper.sv:46]: indicate bitwidth

File: tb/core/Makefile [hunk 9]
```diff
@@ -371,7 +369,7 @@ riscv_tests/rv32um/%.o: riscv_tests/rv32um/%.S riscv_tests/riscv_test.h \
 
 riscv_tests/rv32uc/%.o: riscv_tests/rv32uc/%.S riscv_tests/riscv_test.h \
 			riscv_tests/macros/scalar/test_macros.h
-	$(RISCV_EXE_PREFIX)gcc -c -march=rv32im -g -o $@ \
+	$(RISCV_EXE_PREFIX)gcc -c -march=rv32imc -g -o $@ \
```
Comment (bluewww) [inline tb/core/Makefile:372]: good

File: tb/core/Makefile [hunk 10]
```diff
@@ -83,15 +83,13 @@ RISCV_EXE_PREFIX         = $(RISCV)/bin/riscv32-unknown-elf-
 
 # firmware vars
 FIRMWARE                 = firmware/
-RISCV_TEST_INCLUDES      = -Iriscv_tests/ -Iriscv_tests/macros/scalar \
-				-Iriscv_tests/rv64ui -Iriscv_tests/rv64um -Iriscv_tests/rv64ua
```
Comment (bluewww) [inline tb/core/Makefile:87]: Revert this too, see comment below

File: tb/dm/Makefile [hunk 11]
```diff
@@ -30,7 +30,7 @@ VLOG_FLAGS		= -pedanticerrors -suppress 2577 -suppress 2583
 VLOG_LOG                = vloggy
 
 VOPT			= vopt-$(VVERSION)
-VOPT_FLAGS		= -debugdb -fsmdebug -pedanticerrors #=mnprft
+VOPT_FLAGS		= -debugdb -fsmdebug +acc -pedanticerrors #=mnprft
```
Comment (bluewww) [inline tb/dm/Makefile:33]: Ok this makes sense otherwise the defaults would be really useless.

File: tb/dm/tb_test_env.sv [hunk 12]
```diff
@@ -138,35 +140,29 @@ module tb_test_env
...
         .apu_master_operands_o  (                       ),
         .apu_master_op_o        (                       ),
         .apu_master_type_o      (                       ),
         .apu_master_flags_o     (                       ),
-        .apu_master_valid_i     (                       ),
-        .apu_master_result_i    (                       ),
-        .apu_master_flags_i     (                       ),
-
+        .apu_master_valid_i     ( 1'b0                  ),
+        .apu_master_result_i    (  'b0                  ),
```
Comment (bluewww) [inline tb/dm/tb_test_env.sv:149]: indicate bitwidth. Should stay consistent in this repo.

File: tb/dm/tb_top_verilator.sv [hunk 13]
```diff
@@ -136,15 +139,14 @@ module tb_top_verilator #(
...
         .apu_master_operands_o  (                       ),
         .apu_master_op_o        (                       ),
         .apu_master_type_o      (                       ),
         .apu_master_flags_o     (                       ),
-        .apu_master_valid_i     (                       ),
-        .apu_master_result_i    (                       ),
-        .apu_master_flags_i     (                       ),
-
+        .apu_master_valid_i     ( 1'b0                  ),
+        .apu_master_result_i    (  'b0                  ),
```
Comment (bluewww) [inline tb/dm/tb_top_verilator.sv:148]: indicate bitwidth

File: tb/tb_riscv/riscv_random_stall.sv [hunk 14]
```diff
@@ -86,7 +87,7 @@ endclass : rand_data_cycles
 
 mailbox #(stall_mem_t) core_reqs          = new (4);
 mailbox #(stall_mem_t) core_resps         = new (4);
-mailbox core_resps_granted = new (4);
+mailbox #(logic)       core_resps_granted = new (4);
```
Comment (bluewww) [inline tb/tb_riscv/riscv_random_stall.sv:90]: Revert this if you don't want to break Xcelium. It looks like a bug but it actually fixes a bug. Maybe a comment is in order.

See b54d51a031f4a62a0e3c1a8aecb1ab469c2a652f

File: tb/verilator-model/top.sv [hunk 15]
```diff
@@ -96,37 +97,30 @@ module top
...
       .apu_master_operands_o  (                       ),
       .apu_master_op_o        (                       ),
       .apu_master_type_o      (                       ),
       .apu_master_flags_o     (                       ),
-      .apu_master_valid_i     (                       ),
-      .apu_master_result_i    (                       ),
-      .apu_master_flags_i     (                       ),
-
+      .apu_master_valid_i     ( 1'b0                  ),
+      .apu_master_result_i    (  'b0                  ),
```
Comment (bluewww) [inline tb/verilator-model/top.sv:106]: indicate bitwidth

File: tb/core/mm_ram.sv (hunk 6 above)
Comment (Silabs-ArjanB) [reply tb/core/mm_ram.sv:718]: The connections to riscv_random_stall were actually generating port mismatch warnings and they have been fixed by above change. The key is that I also added an ADDR_WIDTH parameter to riscv_random_stall so that it can be used in different context with potentially different address with requirements (so that we can make this warning free in every context).

File: tb/core/riscv_tests/rv32uc/rvc.S (hunk 7 above)
Comment (Silabs-ArjanB) [reply tb/core/riscv_tests/rv32uc/rvc.S:7]: The ./rv64uc/rvc.S and ./rv32uc/rvc.S file are not the same. Even on the master branch this test case led to a gcc internal error, so the starting point was a repos in which test cases were not compiling correctly.

I looked at what this test looked like roughly half a year ago and I re-applied the same change to get it to work. (This might be related to this issue: https://github.com/riscv/riscv-gnu-toolchain/issues/445)

   diff ./rv64uc/rvc.S ./rv32uc/rvc.S
   10a11,12
   > #undef RVTEST_RV64U
   > #define RVTEST_RV64U RVTEST_RV32U
   16,18d17
   <   // https://github.com/riscv/riscv-gnu-toolchain/issues/445
   <   // mixed modes should turn of relaxation (?)
   <   .option norelax

If we keep the test as it was the entire test suite will not run.

File: tb/core/riscv_wrapper.sv (hunk 8 above)
Comment (Silabs-ArjanB) [reply tb/core/riscv_wrapper.sv:46]: Okay, will do

File: tb/core/Makefile (hunk 10 above)
Comment (Silabs-ArjanB) [reply tb/core/Makefile:87]: Agreed this should be reverted (which has been done now). The rv64* includes are however not actually being used yet as I am still not using the originally intended rvc.S (see discussion above). Once we get the rvc.S issue solved, then at least RISCV_TEST_INCLUDES will be already have the correct value.

File: tb/dm/tb_test_env.sv (hunk 12 above)
Comment (Silabs-ArjanB) [reply tb/dm/tb_test_env.sv:149]: Fixed (and made similar fixes in other places)

File: tb/dm/tb_top_verilator.sv (hunk 13 above)
Comment (Silabs-ArjanB) [reply tb/dm/tb_top_verilator.sv:148]: Done

File: tb/verilator-model/top.sv (hunk 15 above)
Comment (Silabs-ArjanB) [reply tb/verilator-model/top.sv:106]: done

**Comment by davideschiavone** [comment]
Actually, these changes in the RISCV_TESTS suite should be done in the core-verif repository, the only purpose of this "testbench" should be make hello world pass. So it's not wasted work as can be ported to core-verif, but probably we can ignore the them here

File: tb/core/riscv_tests/rv32uc/rvc.S (hunk 7 above)
Comment (bluewww) [reply tb/core/riscv_tests/rv32uc/rvc.S:7]: This might also be a `pulp-riscv-gnu-toolchain` bug. I believe that the test worked with `riscv-gnu-toolchain` but I would have to check again.

**Comment by Silabs-ArjanB** [comment]
Only the assembly/makefile changes seem to have been merged. Many of the proposed RTL and test bench/case updates have not yet been merged to master. Will they be merged later?

For example riscv_core still has the SHARED_* parameters and the Zfinx and DM_HaltAddress parameters according to their old naming style.

Currently (after above merge to master) we are still having warnings when compiling the RTL code, e.g.:

# ** Warning: (vsim-3015) [PCDPC] - Port size (32) does not match connection size (22) for port 'addr_mem_o'. The port definition is at: ../../tb/tb_riscv/riscv_random_stall.sv(52).
#    Time: 0 ns  Iteration: 0  Instance: /tb_top/riscv_wrapper_i/ram_i/instr_random_stalls File: mm_ram.sv Line: 702
# ** Warning: (vsim-3015) [PCDPC] - Port size (32) does not match connection size (22) for port 'addr_mem_o'. The port definition is at: ../../tb/tb_riscv/riscv_random_stall.sv(52).
#    Time: 0 ns  Iteration: 0  Instance: /tb_top/riscv_wrapper_i/ram_i/data_random_stalls File: mm_ram.sv Line: 738

** Warning: tb_test_env.sv(113): (vopt-2685) [TFMPC] - Too few port connections for 'riscv_core_i'.  Expected 46, found 45.
** Warning: tb_test_env.sv(113): (vopt-2718) [TFMPC] - Missing connection for port 'data_atop_o'.
